crates/pglt_lexer/src/diagnostics.rs

```diff
@@ -0,0 +1,63 @@
+use pglt_diagnostics::{Diagnostic, MessageAndDescription};
+use text_size::TextRange;
+
+/// A specialized diagnostic for scan errors.
+///
+/// Scan diagnostics are always **fatal errors**.
+#[derive(Clone, Debug, Diagnostic, PartialEq)]
+#[diagnostic(category = "syntax", severity = Fatal)]
+pub struct ScanError {
+    /// The location where the error is occurred
+    #[location(span)]
+    span: Option<TextRange>,
+    #[message]
+    #[description]
+    pub message: MessageAndDescription,
+}
+
+impl ScanError {
+    pub fn from_pg_query_err(err: pg_query::Error, input: &str) -> Vec<Self> {
+        let err_msg = err.to_string();
+        let re = regex::Regex::new(r#"at or near "(.*?)""#).unwrap();
+        let mut diagnostics = Vec::new();
+
+        for captures in re.captures_iter(&err_msg) {
+            if let Some(matched) = captures.get(1) {
+                let search_term = matched.as_str();
+                for (idx, _) in input.match_indices(search_term) {
```

Comment:
what happens if we have multiple matches in an input? Is that possible?

Reply:
Yes, we return two diagnostics. This is what I meant with my comment about possibly incorrect diagnostics.

Reply:
ahhh okay mh yes that's correct but I don't think it's an issue since this is coming from Postgres and even if there are multiple occurrences the first will be the correct one.